modules/lvgl/lvgl_display.c

```diff
@@ -65,50 +66,46 @@ void lvgl_rounder_cb(lv_disp_drv_t *disp_drv, lv_area_t *area)
 #define lvgl_rounder_cb NULL
 #endif
 
-int set_lvgl_rendering_cb(lv_disp_drv_t *disp_drv)
+int set_lvgl_rendering_cb(lv_display_t *display)
 {
 	int err = 0;
-	struct lvgl_disp_data *data = (struct lvgl_disp_data *)disp_drv->user_data;
+	struct lvgl_disp_data *data = (struct lvgl_disp_data *)lv_display_get_user_data(display);
 
 #ifdef CONFIG_LV_Z_FLUSH_THREAD
-	disp_drv->wait_cb = lvgl_wait_cb;
+	lv_display_set_flush_wait_cb(display, lvgl_wait_cb);
 #endif
 
 	switch (data->cap.current_pixel_format) {
 	case PIXEL_FORMAT_ARGB_8888:
-		disp_drv->flush_cb = lvgl_flush_cb_32bit;
-		disp_drv->rounder_cb = lvgl_rounder_cb;
-#ifdef CONFIG_LV_COLOR_DEPTH_32
-		disp_drv->set_px_cb = NULL;
-#else
-		disp_drv->set_px_cb = lvgl_set_px_cb_32bit;
-#endif
+		lv_display_set_color_format(display, LV_COLOR_FORMAT_ARGB8888);
+		lv_display_set_flush_cb(display, lvgl_flush_cb_32bit);
+		lv_display_add_event_cb(display, lvgl_rounder_cb, LV_EVENT_INVALIDATE_AREA,
+					display);
 		break;
 	case PIXEL_FORMAT_RGB_888:
-		disp_drv->flush_cb = lvgl_flush_cb_24bit;
-		disp_drv->rounder_cb = lvgl_rounder_cb;
-		disp_drv->set_px_cb = lvgl_set_px_cb_24bit;
+		lv_display_set_color_format(display, LV_COLOR_FORMAT_RGB888);
+		lv_display_set_flush_cb(display, lvgl_flush_cb_24bit);
+		lv_display_add_event_cb(display, lvgl_rounder_cb, LV_EVENT_INVALIDATE_AREA,
+					display);
 		break;
 	case PIXEL_FORMAT_RGB_565:
 	case PIXEL_FORMAT_BGR_565:
-		disp_drv->flush_cb = lvgl_flush_cb_16bit;
-		disp_drv->rounder_cb = lvgl_rounder_cb;
-#ifdef CONFIG_LV_COLOR_DEPTH_16
-		disp_drv->set_px_cb = NULL;
-#else
-		disp_drv->set_px_cb = lvgl_set_px_cb_16bit;
-#endif
+		lv_display_set_color_format(display, LV_COLOR_FORMAT_RGB565);
+		lv_display_set_flush_cb(display, lvgl_flush_cb_16bit);
+		lv_display_add_event_cb(display, lvgl_rounder_cb, LV_EVENT_INVALIDATE_AREA,
+					display);
 		break;
 	case PIXEL_FORMAT_MONO01:
 	case PIXEL_FORMAT_MONO10:
```

Comment:
This causes issues in the `tests/lib/gui/lvgl` for the mono test case `libraries.gui.lvgl.mono`

`west build -p -b native_sim/native tests/lib/gui/lvgl -T libraries.gui.lvgl.mono`

Reason is that there is a discrepency between the usage of the LV_COLOR_DEPTH kconfigs of LVGL and the pixel ofrmat that one assigns to the display. In the testcase we set `CONFIG_LV_COLOR_DEPTH_1` but this does not change the actual pixel format of the display. Therefore with the new optimization for monochrome displays (1 bit per pixel instead of a whole byte) this fails internal buffer length checks of LVGL.  Raised this PR https://github.com/zephyrproject-rtos/zephyr/pull/81263 so we can add overlays for the tests to ensure that the code paths are run as expected.

Reply:
Essentially this is what the test code now does:
https://github.com/zephyrproject-rtos/zephyr/blob/f9981e34b4b4d3254e774986082b097fbf57aafc/tests/lib/gui/lvgl/src/main.c#L118

Before calling the init manually the display pixel format is set. Static allocation falls flat here, because your buffer needs to accomodate for a possible runtime switch. This means if you set `LV_Z_BITS_PER_PIXEL` high and switch the pixel format to a smaller one you waste some memory on your static draw buffers. But there is little you can do about.

Reply:
Sure, that makes sense about static allocation. I think what I'm confused about is if is possible to generically set the display pixel format before `lvgl_init` runs, or if we need to use this workaround to explicitly delay init.

IE could we modify our LVGL porting code so that the display pixel format is set before we call `lvgl_init`, and avoid the assertion?

Reply:
Not sure if I understand the suggested change.

> IE could we modify our LVGL porting code so that the display pixel format is set before we call lvgl_init, and avoid the assertion?

Just to make sure that we are on the same page 😄 :
The display driver is initialized with the `POST_KERNEL` init priority. Most in-tree drivers have a default current `pixel_format` set after init is run. For some device tree bindings the initial `pixel_format` is configurable via device tree.
After this the LVGL init code is run inside the `SYS_INIT` with `APPLICATION` init priority.

So there are two cases:

- The display has the "correct" pixel format set after init. In this case static allocation and everything else will work just fine. The `pixel_format` is "visible" to the LVGL init code during `SYS_INIT`
- The display has a "wrong" pixel format configured. In this case one would need to set `LV_Z_AUTO_INIT=n` and somewhere within application code do the following:
```
display_set_pixel_format(dev, PIXEL_FORMAT_MONO01);
lvgl_init();
```
This will work with the dynamic allocation just fine, with the static one there is the potential that you have to over-allocate.

Does your question mean that you want the LVGL init code to set the displays `pixel_format`? So having a Kconfig property like `LV_Z_DISPLAY_PIXEL_FORMAT` or the like and set the pixel format based on that during the init?

Reply:
> Does your question mean that you want the LVGL init code to set the displays pixel_format? So having a Kconfig property like LV_Z_DISPLAY_PIXEL_FORMAT or the like and set the pixel format based on that during the init?

Honestly, I think this *might* make sense. LVGL already has a good idea of the pixel format it expects from the display based on the choice of `CONFIG_LV_COLOR_DEPTH`, right?

- `LV_COLOR_DEPTH_32`: `PIXEL_FORMAT_ARGB_8888`
- `LV_COLOR_DEPTH_16`: `PIXEL_FORMAT_RGB_565` if `LV_COLOR_16_SWAP == y`, otherwise `PIXEL_FORMAT_BGR_565`
- `LV_COLOR_DEPTH_1`: check if pixel format is `PIXEL_FORMAT_MONO01` or `PIXEL_FORMAT_MONO_10`, if not set the format to `MONO01`.

The `MONO10` versus `MONO01` does leave some ambiguity, which might be an issue. Otherwise though I think this would be useful- LVGL would be able to detect during init if a display could not support the chosen pixel format

Reply:
Hmm might make sense for an enhancement, in the future. With the current changeset we are not painting ourselves into a corner. The delayed init was request by some people using deferred-init with their displays so it's needed either way. I want to get this one merged first and then think about enhancements. LVGL also supports multiple screens so a refactoring of sorts will be needed for that.